src/wallet/mod.rs

```diff
@@ -253,32 +261,91 @@ where
             .map_err(|_| Error::ScriptDoesntHaveAddressForm)
     }
 
-    // Return the the last previously derived address for `keychain` if it has not been used in a
-    // received transaction. Otherwise return a new address using [`Wallet::get_new_address`].
-    fn get_unused_address(&self, keychain: KeychainKind) -> Result<AddressInfo, Error> {
+    // Return vector of unused address indexes
+    fn get_unused_key_indexes(&self, keychain: KeychainKind) -> Result<Vec<u32>, Error> {
         let current_index = self.fetch_index(keychain)?;
 
-        let derived_key = self
-            .get_descriptor_for_keychain(keychain)
-            .as_derived(current_index, &self.secp);
-
-        let script_pubkey = derived_key.script_pubkey();
+        let derived_keys: Vec<Descriptor<DerivedDescriptorKey>> = (0..current_index + 1)
+            .map(|i| {
+                self.get_descriptor_for_keychain(keychain)
+                    .as_derived(i, &self.secp)
+            })
+            .collect();
 
-        let found_used = self
-            .list_transactions(true)?
+        let unused_key_indexes: Vec<u32> = derived_keys
             .iter()
-            .flat_map(|tx_details| tx_details.transaction.as_ref())
-            .flat_map(|tx| tx.output.iter())
-            .any(|o| o.script_pubkey == script_pubkey);
+            .cloned()
+            .enumerate()
+            .filter_map(|(i, key)| {
+                if !self
+                    .list_transactions(true)
+                    .expect("getting transaction list")
+                    .iter()
+                    .flat_map(|tx_details| tx_details.transaction.as_ref())
+                    .flat_map(|tx| tx.output.iter())
+                    .any(|o| o.script_pubkey == key.script_pubkey())
+                {
+                    Some(i as u32)
+                } else {
+                    None
+                }
+            })
+            .collect();
+
+        Ok(unused_key_indexes)
+    }
 
-        if found_used {
+    // Return the the last previously derived address for `keychain` if it has not been used in a
+    // received transaction. Otherwise return a new address using [`Wallet::get_new_address`].
+    fn get_last_unused_address(&self, keychain: KeychainKind) -> Result<AddressInfo, Error> {
```

Comment:
I am thinking if we already got `get_first_unused_address` what we can use `get_last_unused_address` for?? Isn't it always better to use the first unused in all situations?

Reply:
Not necessarily I think. `last_unused_address` basically means the last address you gave out (unless it has been used). It's useful when you need a new address but you can't remember if you used the last one the app gave out to you.

Reply:
But even in the case if I don't remember i I used the last one I fetched, won't I still be wanting the first unused?? Won't that always ensure that address use gaps are minimized over the natural usage of the wallet?? 

If I ask last unused there might be other addresses too I could have used, just didn't knew.. So those gap remains until someone calls first unused again..

Reply:
I don't have a strong rationale for `last_unused`'s existence tbh. We could remove it without complaint from me :)

Reply:
Yep I have changed LastUnused here to 'go back' and find the last unused address, rather than just checking only the address at current index and returning new otherwise. If someone gets addresses A1 and A2, and then uses A2. Do we want `LastUnused` to return new or A1?

This behavior made more sense to me for `LastUnused` address, not something like `CurrentAddressIfUnused`. Do you think this behavior is unexpected or too complex?

I agree that `FirstUnused` is better for minimizing address gaps, and yeah not exactly sure of when `LastUnused` would be preferable. Other than lazily getting the address you're about to use in something like
>when you need a new address but you can't remember if you used the last one the app gave out to you.

Reply:
Are you thinking we leave the current LastUnused as it currently exists? or rename to `CurrentAddressIfUnused` and implement `LastUnused` to mean "give me the last address that hasn't been used, otherwise new"